File: C++/SpiralMatrix.cpp
#include<iostream>
#include<vector>
#include<string>
#include<string.h>
#include<math.h>
using namespace std;
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int> > &matrix) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
		vector<int> ret;
		int len1 = matrix.size();
		if(!len1)
			return ret;
		int len2 = matrix[0].size();
		int small = min(len1, len2);
		int big = max(len1, len2);
		int level = 0;
		for(; level < small/2; level++)
		{
			for(int i = level; i < len2-level; i++)
				ret.push_back(matrix[level][i]);
			for(int i = level+1; i < len1 - level; i++)
				ret.push_back(matrix[i][len2 - level - 1]);
			for(int i = len2 - level - 2; i >= level; i--)
				ret.push_back(matrix[len1 - level - 1][i]);
			for(int i = len1 - level - 2; i >= level+1; i--)
				ret.push_back(matrix[i][level]);
		}
		if(small % 2)
		{
			if(big == len1)
			{
				for(int i = level; i < len1 - level; i++)
					ret.push_back(matrix[i][level]);
			}
			else
			{
				for(int i = level; i < len2 - level; i++)
					ret.push_back(matrix[level][i]);
			}
		}
		return ret;
    }
};
int main()
{
	int tmp[3][2] = {{2,5},{8,4},{0,-1}};
	vector<vector<int>> matrix(3);
	for(int i = 0; i < 3; i++)
	{
		matrix[i] = vector<int>(tmp[i], tmp[i] + 2);
	}
	Solution sol;
	sol.spiralOrder(matrix);
	getchar();
}

File: JAVA/SpiralMatrix.java
import java.util.ArrayList;

class Solution {
	ArrayList<Integer> ret = new ArrayList<Integer>();

	public ArrayList<Integer> spiralOrder(int[][] matrix) {
		// Start typing your Java solution below
		// DO NOT write main() function
		ret.clear();
		if (matrix == null || matrix.length == 0)
			return ret;
		int m = matrix.length;
		int n = matrix[0].length;

		int level = 0;
		for (; level < m / 2 && level < n / 2; level++) {
			for (int i = level; i < n - level; i++) {
				ret.add(matrix[level][i]);
			}
			for (int i = level + 1; i < m - level; i++) {
				ret.add(matrix[i][n - level - 1]);
			}
			for (int i = n - level - 2; i >= level; i--) {
				ret.add(matrix[m - level - 1][i]);
			}
			for (int i = m - level - 2; i >= level + 1; i--) {
				ret.add(matrix[i][level]);
			}
		}
		if (m % 2 == 1 || n % 2 == 1) {
			for (int i = level; i < m - level; i++) {
				for(int j = level; j < n - level; j++)
				ret.add(matrix[i][j]);
			}
		}
		return ret;
	}
}

public class SpiralMatrix {

	/**
	 * @param args
	 */
	public static void main(String[] args) {
		int[][] matrix = { { 2, 3, 4 }, { 5, 6, 7 }, { 8, 9, 10 },
				{ 11, 12, 13 } };
		Solution sol = new Solution();
		System.out.println(sol.spiralOrder(matrix));
	}

}
